### ISSUE TITLE
Cannot type accented letters

### ISSUE THREAD

**Issue by raspbeguy** [davidgiven/wordgrinder#37]
Hello,
I am unable to type any accented letter like `é`,`è`,`ç`,`à`... As a french user it is very annoying for me. When I type any accented letter, I cannot type anything else, even regular letters, until I press backspace.

I'm using urxvt on archlinux, and the problem is present in the AUR release version (0.6.0) and in the git version.

Am I missing something ?

**Comment by davidgiven** [comment]
I don't believe that's my fault --- urxvt is handling all the keyboard input, and I just enter into your document whatever character I get from it.

Can you type accented characters on the command line?

**Comment by raspbeguy** [comment]
yes, I can type those characters in the terminal and in any other program like vim.
I just tried on another computer and with xterm, it's the same problem.

**Comment by raspbeguy** [comment]
It it can be of any help, I am also unable to type other characters like `£`, `µ`, `§`, `²`. I cannot paste them into wordgrinder either.

**Comment by davidgiven** [comment]
Okay, that does sound like my fault, then. But I use urxvt myself, on Debian, and it all works fine there...

Could you run the `locale` command and send me the output, please?

**Comment by raspbeguy** [comment]
Sure, here it is:

```LANG=fr_FR.utf8
LC_CTYPE="fr_FR.utf8"
LC_NUMERIC="fr_FR.utf8"
LC_TIME="fr_FR.utf8"
LC_COLLATE="fr_FR.utf8"
LC_MONETARY="fr_FR.utf8"
LC_MESSAGES="fr_FR.utf8"
LC_PAPER="fr_FR.utf8"
LC_NAME="fr_FR.utf8"
LC_ADDRESS="fr_FR.utf8"
LC_TELEPHONE="fr_FR.utf8"
LC_MEASUREMENT="fr_FR.utf8"
LC_IDENTIFICATION="fr_FR.utf8"
LC_ALL=
```

I wrote a txt in vim containing the problematic characters and then I tried to import it in wordgrinder, then weird shit happens... I screencaptured it [here](https://hashtagueule.fr/fichiers/capture.mp4) (the video seems corrupted but you can play it just fine for instance in mpv)

**Comment by raspbeguy** [comment]
Is there a way I could log events or enable debug mode ?

**Comment by davidgiven** [comment]
Not really --- I tend to just bodge in that sort of thing when I need it. Sorry, that's not very helpful.

I have a suspicion that the ncurses library hasn't been built with UTF-8 support. Could you do:

    $ ldd /usr/bin/wordgrinder

...and tell me what it produces? Also, you might want to try `xwordgrinder`, which is the X version --- if that doesn't work too, something very odd is happening.

Also, if you could tell me what Arch version you're using, I'll try to duplicate your setup in a VM.

**Comment by raspbeguy** [comment]
```
ldd /usr/bin/wordgrinder
	linux-vdso.so.1 (0x00007ffff7964000)    
	liblua5.2.so.5.2 => /usr/lib/liblua5.2.so.5.2 (0x00007f6986c3a000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007f6986a23000)
	libncursesw.so.6 => /usr/lib/libncursesw.so.6 (0x00007f69867b5000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f6986411000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f698620d000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f6985efa000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f6986e6d000)
```
I cannot type accented characters on `xwordgrinder` either (which is not installed by the AUR package, I had to clone manually your repository).

Arch being a rolling release, there isn't actually a version for it.

**Comment by raspbeguy** [comment]
I made some tests:
* I installed wordgrinder on a brand new Arch installation on a VM with Gnome (I can give you the image), it doesn't work better.
* I installed wordgrinder on a Ubuntu 17.04, and here it works. I noted also that worgrinder is far prettier on Ubuntu (menus with borders, top and bottom of the page...), so it can show a possible difference in ncurses libraries.

**Comment by davidgiven** [comment]
The image would be really helpful, thanks!

Yes, I noticed that lack of menus in your video. I put that down to the font you were using not having the unicode box drawing characters --- it displayed accented characters without any trouble, so I assumed unicode was working. It should, hopefully, look better with a different font?

**Comment by raspbeguy** [comment]
Nope, tried to run it with Ubuntu font, same problem.

**Comment by davidgiven** [comment]
That's very interesting. It suggests that there's an ncurses problem. The ldd log you showed me did say it was the wide-character version of ncurses (ncursesw). Maybe it's broken somehow? I did find some Arch bugs about the ncurses library --- but they date from 2008...

**Comment by raspbeguy** [comment]
The image is 20 GB big (I know, I could have made it smaller)... I can send it without problem, are you still wanting to have it ?

**Comment by davidgiven** [comment]
Hm. I wasn't expecting it to be quite *that* big. How difficult is your setup to reproduce?

**Comment by raspbeguy** [comment]
Well, I'll see, but tomorrow, now time to sleep :)
meanwhile, if you want it anyway, here is the link https://home.gugod.fr/files/triomphe.qcow2

**Comment by raspbeguy** [comment]
I didn't have the time yet to create another VM, but I'm sure it can be easily reproduced : Archlinux, locale `fr_FR.UTF-8`.

**Comment by davidgiven** [comment]
I've downloaded the VM, but can't make it boot --- on startup it says it's launched Gnome, but nothing happens (it remains at the bootup console). It probably doesn't like my qemu setup. What virtual machine were you using?

**Comment by raspbeguy** [comment]
I am using KVM on libvirt.

**Comment by raspbeguy** [comment]
btw, user is raspbeguy, password is wordgrinder

**Comment by davidgiven** [comment]
Well, I got the image working... turns out that the main problem was Wayland; disabling that made X start, but then it was refreshing at about 0.5 fps (no exaggeration). I'd forgotten how much Gnome annoys me (really, they hide the logout button?). Eventually I realised that I could reproduce the problem without needing to start X.

But I reproduced it! And diagnosed it.

Turns out that Arch doesn't have the C.UTF-8 locale, which WordGrinder uses internally. So it's falling back to the C locale, which doesn't support UTF-8, which means very bad things happen when UTF-8 characters are processed.

If you want a quick fix, open `src/c/main.c` and change the second argument of `setlocale()` from `C.UTF-8` to `en_US.UTF-8`. Things will mostly work but I can't guarantee that weird stuff won't happen.

Trouble is, a _proper_ fix is harder. I need UTF-8, but I can't use a language-specific locale (because odd things happen with stuff like sort order). If I can't use `C.UTF-8` I'm not sure what I can use. `en_US.UTF-8` isn't suitable for general use because it, e.g., sorts upper and lower case together.

I'll have to think about this one.

**Comment by raspbeguy** [comment]
Ok, thanks for your investigation!
Does it change anything if I replace `en_US.UTF-8` with `fr_FR.UTF-8` ? I don't really know what language specific locale change at all in this context.

**Comment by davidgiven** [comment]
The tl;dr is: your version has to run in English (because it's relying on an error message we get from Lua being in English to detect an error condition).

However, I've just updated the version in git not to do this any more. So if you check this out, then edit the locale in `src/c/main.c` to `""` (which tells it to use the locale of your system), then it may work.

Important note: I can't guarantee that it _will_ work when using a non-`C` locale. Please let me know if you see any odd behaviour.

Also, WordGrinder isn't localised in any way, so the behaviour when using `fr_FR.UTF-8` should be identical to `en_US.UTF-8`...

**Comment by davidgiven** [comment]
Have you been using WordGrinder with either the `en_US.UTF-8` or `fr_FR.UTF-8` locales? Have you observed any odd behaviour?

Having thought about this I think that the least bad thing to do is to try several locales --- `C.UTF-8` first, and then various common locales, and then the default system locale, until one works. If no UTF-8 locale could be found, then it just runs without UTF-8. I don't think there's any better thing I can do here.

But all my test systems have C.UTF-8, so it'd be good to know whether a language-specific locale actually works.

**Comment by raspbeguy** [comment]
Ok sorry for the delay.
I can confirm that it works now. Thanks for your work !

**Comment by davidgiven** [comment]
Excellent, thanks.

Yes, I agree. There's one feature I need to think about and implement (a UI option to allow first-line-indented-no-space-between-paragraphs mode) and then it should be ready to go. Hopefully, soon...
